base/tuple.jl

```diff
@@ -20,7 +20,7 @@ endof(t::Tuple) = length(t)
 size(t::Tuple, d) = d==1 ? length(t) : throw(ArgumentError("invalid tuple dimension $d"))
 getindex(t::Tuple, i::Int) = getfield(t, i)
 getindex(t::Tuple, i::Real) = getfield(t, convert(Int, i))
-getindex(t::Tuple, r::AbstractArray{<:Any,1}) = tuple([t[ri] for ri in r]...)
+getindex(t::Tuple, r::AbstractArray{<:Any,1}) = (n = length(eachindex(r)); ntuple(i->t[r[i]], n))
```

Comment:
> I feel like we are slowly moving to support non traditional indexing, to handle things like, e.g., OffsetArrays. That's why I wrote it like this.

Exactly :). The proposed definition indexes `r` from `1:n`, as `i` iterates over `1:n` in `ntuple`. So if indexing `r` from `1:n` is not correct, the proposed definition will not be correct.

> This is already non-inferable so it doesn't hurt changing it to make it faster.

Consider
```julia
julia> foo(t::Tuple, r::AbstractArray{<:Any,1}) = tuple([t[ri] for ri in r]...)
foo (generic function with 1 method)

julia> bar(t::Tuple, r::AbstractArray{<:Any,1}) = (n = length(eachindex(r)); ntuple(i->t[r[i]], n))
bar (generic function with 1 method)

julia> @code_warntype foo((1,2,3), [1])
Variables:
  #self#::#foo
  t::Tuple{Int64,Int64,Int64}
  r::Array{Int64,1}
  #1::##1#2{Tuple{Int64,Int64,Int64}}

Body:
  begin
      #1::##1#2{Tuple{Int64,Int64,Int64}} = $(Expr(:new, :($(QuoteNode(##1#2{Tuple{Int64,Int64,Int64}}))), :(t)))
      SSAValue(0) = #1::##1#2{Tuple{Int64,Int64,Int64}}
      SSAValue(1) = $(Expr(:new, :($(QuoteNode(Base.Generator{Array{Int64,1},##1#2{Tuple{Int64,Int64,Int64}}}))), SSAValue(0), :(r)))
      SSAValue(2) = $(Expr(:invoke, MethodInstance for collect(::Base.Generator{Array{Int64,1},##1#2{Tuple{Int64,Int64,Int64}}}), :(Base.collect), SSAValue(1)))
      return (Core._apply)(Main.tuple, SSAValue(2))::Tuple{Vararg{Int64,N} where N}
  end::julia> foo(t::Tuple, r::AbstractArray{<:Any,1}) = tuple([t[ri] for ri in r]...)
foo (generic function with 1 method)

julia> bar(t::Tuple, r::AbstractArray{<:Any,1}) = (n = length(eachindex(r)); ntuple(i->t[r[i]], n))
bar (generic function with 1 method)

julia> @code_warntype foo((1,2,3), [1])
Variables:
  #self#::#foo
  t::Tuple{Int64,Int64,Int64}
  r::Array{Int64,1}
  #1::##1#2{Tuple{Int64,Int64,Int64}}

Body:
  begin
      #1::##1#2{Tuple{Int64,Int64,Int64}} = $(Expr(:new, :($(QuoteNode(##1#2{Tuple{Int64,Int64,Int64}}))), :(t)))
      SSAValue(0) = #1::##1#2{Tuple{Int64,Int64,Int64}}
      SSAValue(1) = $(Expr(:new, :($(QuoteNode(Base.Generator{Array{Int64,1},##1#2{Tuple{Int64,Int64,Int64}}}))), SSAValue(0), :(r)))
      SSAValue(2) = $(Expr(:invoke, MethodInstance for collect(::Base.Generator{Array{Int64,1},##1#2{Tuple{Int64,Int64,Int64}}}), :(Base.collect), SSAValue(1)))
      return (Core._apply)(Main.tuple, SSAValue(2))::Tuple{Vararg{Int64,N} where N}
  end::Tuple{Vararg{Int64,N} where N}

julia> @code_warntype bar((1,2,3), [1])
Variables:
  #self#::#bar
  t::Tuple{Int64,Int64,Int64}
  r::Array{Int64,1}
  n::Int64
  #3::##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}}

Body:
  begin
      $(Expr(:inbounds, false))
      # meta: location abstractarray.jl eachindex 744
      # meta: location abstractarray.jl indices1 63
      # meta: location abstractarray.jl indices 56
      SSAValue(3) = (Base.arraysize)(r::Array{Int64,1}, 1)::Int64
      # meta: location tuple.jl map 124
      SSAValue(2) = SSAValue(3)
      # meta: pop location
      # meta: pop location
      # meta: pop location
      # meta: pop location
      $(Expr(:inbounds, :pop))
      n::Int64 = (Core.typeassert)((Base.select_value)((Base.slt_int)(SSAValue(2), 0)::Bool, 0, SSAValue(2))::Int64, Int64)::Int64
      #3::##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}} = $(Expr(:new, :($(QuoteNode(##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}}))), :(t), :(r)))
      SSAValue(0) = #3::##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}}
      return $(Expr(:invoke, MethodInstance for ntuple(::##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}}, ::Int64), :(Main.ntuple), SSAValue(0), :(n)))
  end::Tuple

julia>

julia> @code_warntype bar((1,2,3), [1])
Variables:
  #self#::#bar
  t::Tuple{Int64,Int64,Int64}
  r::Array{Int64,1}
  n::Int64
  #3::##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}}

Body:
  begin
      $(Expr(:inbounds, false))
      # meta: location abstractarray.jl eachindex 744
      # meta: location abstractarray.jl indices1 63
      # meta: location abstractarray.jl indices 56
      SSAValue(3) = (Base.arraysize)(r::Array{Int64,1}, 1)::Int64
      # meta: location tuple.jl map 124
      SSAValue(2) = SSAValue(3)
      # meta: pop location
      # meta: pop location
      # meta: pop location
      # meta: pop location
      $(Expr(:inbounds, :pop))
      n::Int64 = (Core.typeassert)((Base.select_value)((Base.slt_int)(SSAValue(2), 0)::Bool, 0, SSAValue(2))::Int64, Int64)::Int64
      #3::##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}} = $(Expr(:new, :($(QuoteNode(##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}}))), :(t), :(r)))
      SSAValue(0) = #3::##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}}
      return $(Expr(:invoke, MethodInstance for ntuple(::##3#4{Tuple{Int64,Int64,Int64},Array{Int64,1}}, ::Int64), :(Main.ntuple), SSAValue(0), :(n)))
  end::Tuple
```
The existing definition yields `Tuple{Vararg{Int64,N} where N}` whereas the propose definition yields only `Tuple`; some information is lost?